server/src/main/java/io/crate/execution/engine/join/HashJoinBatchIterator.java

```diff
@@ -264,46 +243,31 @@ private boolean buildBufferAndMatchRight() {
                 numItems++;
                 int hash = hashBuilderForLeft.applyAsInt(unsafeArrayRow.cells(leftRow));
                 addToBuffer(leftRow, hash);
-                if (numberOfRowsInBuffer == blockSize || circuitBreaker.getFree() < 512 * 1024) {
+                if (numberOfHashGroupsInBuffer == blockSize || circuitBreaker.getFree() < 512 * 1024) {
                     break;
                 }
             }
             leftAverageRowSize = numItems > 0 ? (int) (sum / numItems) : -1;
 
-            if (numberOfRowsInBuffer == 0 && !left.allLoaded()) {
+            if (numberOfHashGroupsInBuffer == 0 && !left.allLoaded()) {
                 return false;
             } else {
                 activeIt = right;
             }
         }
 
         // In case of multiple matches on the left side (duplicate values or hash collisions)
-        if (leftMatchingRowsIterator != null && findMatchingRows()) {
-            if (emitNullValues) {
-                // We found matching rows, therefore we mark the values to emit
-                // non-matching values later with null value pairs
-                if (leftMatchingRows.matched == false) {
-                    leftMatchingRows.matched = true;
-                }
-            }
+        if (matchedHashGroup != null && findMatchingRows()) {
             return true;
         }
 
-        leftMatchingRowsIterator = null;
         while (right.moveNext()) {
             int rightHash = hashBuilderForRight.applyAsInt(right.currentElement());
-            leftMatchingRows = buffer.get(rightHash);
-            if (leftMatchingRows != null) {
-                leftMatchingRowsIterator = leftMatchingRows.items.iterator();
+            matchedHashGroup = buffer.get(rightHash);
+            if (matchedHashGroup != null) {
+                matchedHashGroupIdx = 0;
```

Comment:
Can we revert this and use the `HashGroup.iterator()` as the `leftMatchingRowsIterator`?